test/quantization/ptq/wrappers/fairseq/test_quant_encoder_layer.py

```diff
@@ -0,0 +1,257 @@
+
+        # LayerNorms
+        self.self_attn_layer_norm = nn.LayerNorm(embed_dim)
+        self.final_layer_norm = nn.LayerNorm(embed_dim)
+
+        # Activation
+        self.activation_fn = nn.GELU()
+
+    def forward(self, *args, **kwargs):
+        raise NotImplementedError("Not used in tests; wrapper handles forward.")
+
+
+class _DummySelfAttn(nn.Module):
+    """Self-attention-shaped module providing q/k/v/out_proj linears and fairseq flags."""
+
+    def __init__(self, embed_dim, num_heads, bias=False):
+        super().__init__()
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+        self.self_attention = True
+        self.encoder_decoder_attention = False
+        self.q_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.k_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.v_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.out_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+
+
+class TestQuantFairseqEncoderLayer(unittest.TestCase):
+    def setUp(self):
+        torch.manual_seed(0)
+        self.E = 16  # embed dim
+        self.H = 4  # num heads
+        self.B = 3  # batch size
+        self.T = 5  # sequence length
+
+    def _make_inputs(self, T=None):
+        T = T or self.T
+        x = torch.randn(T, self.B, self.E)
+        return x
+
+    def _build_layer(
+        self, normalize_before=True, return_fc=False, qcfg: Optional[QuantConfig] = None
+    ):
+        fp = DummyFSEncoderLayer(
+            embed_dim=self.E,
+            num_heads=self.H,
+            normalize_before=normalize_before,
+            return_fc=return_fc,
+        )
+        layer = QuantFairseqEncoderLayer(fp, qcfg=qcfg, fp_name="enc0")
+        return layer, fp
+
+    # 1) Forward shape for both pre/post-norm and with/without return_fc
+    def test_forward_shapes_flags(self):
+        for normalize_before in (True, False):
+            for return_fc in (True, False):
+                layer, _ = self._build_layer(normalize_before, return_fc, QuantConfig())
+                x = self._make_inputs()
+                out = layer(x, encoder_padding_mask=None, attn_mask=None)
+                if return_fc:
+                    y, fc = out
+                    self.assertEqual(y.shape, (self.T, self.B, self.E))
+                    self.assertEqual(fc.shape, (self.T, self.B, self.E))
+                else:
+                    y = out
+                    self.assertEqual(y.shape, (self.T, self.B, self.E))
+
+    # 2) Mask handling: boolean causal attn_mask and boolean key_padding_mask
+    def test_mask_handling(self):
+        layer, _ = self._build_layer(True, False, QuantConfig())
+        x = self._make_inputs()
+        # Upper-triangular causal mask [T,S] (True = masked)
+        attn_mask_bool = torch.triu(
+            torch.ones(self.T, self.T, dtype=torch.bool), diagonal=1
+        )
+        # Key padding mask [B,S] (True = pad)
```

Comment:
Q. About abbreviation `T` and `S`, it seems both refers to sequence length but what's difference?

Reply:
`T` refers to target sequence length and `S` refers to source sequence length. 

When it comes to the comments in the codes, `attn_mask` has shape (T, S), where T = query length and S = key length. And, `key_padding_mask` is (B, S), since it marks padding positions in the key sequence for each batch.